### ISSUE TITLE
[bndtools] Automatically add source bundles when debugging

### ISSUE THREAD

**Issue by kriegfrj** [bndtools/bnd#3899]
The debugger seems to the sources are embedded in the main bundle under `OSGI-OPT/src` (as is the case with most bundles built by Bnd).

However, the Eclipse bundles use the "separate bundle for the source" paradigm. I'm trying to implement #3881 and getting a little frustrated by the fact that the debugger can't find them automatically. This is compounded by the fact that Bnd doesn't even automatically download them for you into the Eclipse cache merely by being on `-runbundles`. So to get it to work, I seem to need to:

* add the bundle to `-buildpath` and open the source (force
* run in debug mode
* do "add sources" in the debugger and manually browse to bnd's Eclipse cache folder to find the source bundle that I need and add it.

I think that all of this is something that `AbstractOSGILaunchDelegate` should be able to handle on behalf of the user when it detects that it has been launched in debug mode.

**Comment by pkriens** [comment]
I remember Neil was searching for <bsn>.sources and attached them as source. I also know I implemented this for the maven repos.

**Comment by kriegfrj** [comment]
Yes, it seems to work for the Maven-sourced dependencies. But for BuildpathQuickFixProcessorTest I'm using Eclipse (p2)-sourced dependencies.

And even with the .source bundles, it seems to be able to find them if you have the bundle on the build path. But often you have bundles in `-runbundles` that are not on the `-buildpath`, and it seems the debugger is not finding those.

Antother possibility is that when BJ builds the OSGi Repository version of Eclipse 2018-12, he could package all of the source into the OSGI-OPT/src folder of the main bundle. I think that would work, but it would be a fix for our specific scenario. It'd be nice to have something a bit more generic.

**Comment by pkriens** [comment]
+1 to the idea of adding sources of all -runbundles in the debug launching. As far as I can see, that should be pretty straightforward? That would actually be great.

**Comment by kriegfrj** [comment]
I've had an initial look into this, as the lack of debug sources is really causing me problems when trying to debug the quick fix integration tests. Reporting my findings for feedback, and also for my future self in case I have to move off onto something else.

There is already a crude implementation of this in the form of `bndtools.launch.BndDependencySourceContainer` (`bndtools.core`). It wraps all of the bundles list in `-runbundles` into an `ArchiveSourceContainer` and adds them to a source dependency group called "Bnd Dependencies". This works for bundles that have their source embedded somewhere (eg, `OSGI-OPT/src`). However, for repos like Eclipse 2018-12 which don't have the source embedded but have it in a separate bundle, it is not particularly useful - all it does is slow down the source lookup.

I managed to get a crude implementation of this to work. This involved transplanting `org.bndtools.builder.classpath.BndContainerSourceManager.getSourceBundle()` (in `bndtools.builder`) into `BndDependencySourceContainer`. It sort-of works, but there are a number of issues. One issue is that it searches all repos for a matching source bundle every time you look for the source, which is slightly inefficient. But the biggest problem is that the first time you run it, typically you don't have manny of the source bundles cached locally yet - because `ArchiveSourceContainer` requires a file on the local file system (in the local workspace), it has to download each one before it constructs the corresponding `ArchiveSourceContainer`. Because of the structure of the code, this is currently done sequentially and sometimes on the UI thread - downloading all of the source for an Eclipse launch can pin the UI thread for some time. 

I'm thinking that the best way to do this might be to create a custom `ISourceContainer` implementation that is repository-aware. It could even appear visually different to a regular Archive or External Archive in the Debug Source window, using the Bndtools bundle icon and using the BSN + version descriptor (like the runbundles display in the bnd(run) editor).  Apart from being visually distinctive, the advantages of such an implementation is that it could interact directly with the Repository abstraction rather than via the file system. This would in turn potentially make things possible/easier, like lazy downloading of source bundles - you wouldn't need to actually download the bundle until a source lookup operation happens that requires your particular bundle. It is possible the class lookup index could be leveraged here too.

**Comment by kriegfrj** [comment]
The more I look into this, the more I think that `ArchiveSourceContainer` is not the best choice for wrapping the Bnd dependencies.

My initial implementation didn't work because I used an `ArchiveSourceContainer`. `ArchiveSourceContainer` is used for jars that are part of the Eclipse workspace. In my case, the sources had been successfully downloaded and were sitting in `cnf/cache/stable/Eclipse 2018-12`, but I had to manually refresh that folder in the Eclipse workspace before the archives were added to the workspace model. Moreover, in general the jars will not be part of the workspace (eg, those that are in the Maven cache).
